numba/core/untyped_passes.py

```diff
@@ -1068,116 +1067,145 @@ def get_call_args(init_arg, want):
+                assert isinstance(ty, self._accepted_types)
+                # loop header is spelled ok, now make sure the body
+                # actually contains a getitem
+
+                # find a "getitem"... only looks in the blocks that belong
+                # _solely_ to this literal_unroll (there should not be nested
+                # literal_unroll loops, this is unsupported).
+                tuple_getitem = None
+                for lbli in loop.body:
+                    blk = func_ir.blocks[lbli]
+                    for stmt in blk.body:
+                        if isinstance(stmt, ir.Assign):
+                            if (isinstance(stmt.value, ir.Expr) and
+                                    stmt.value.op == "getitem"):
+                                # check for something like a[i]
+                                if stmt.value.value != arg:
+                                    # that failed, so check for the
+                                    # definition
+                                    dfn = guard(get_definition, func_ir,
+                                                stmt.value.value)
+                                    if dfn is None:
                                         continue
-                                    tuple_getitem = stmt
-                                    break
-                        if tuple_getitem:
-                            break
-                    else:
-                        continue  # no getitem in this loop
-
-                    ui = unroll_info(loop, literal_unroll_call, arg,
-                                     tuple_getitem)
-                    literal_unroll_info[lbl] = ui
-
+                                    try:
+                                        args = getattr(dfn, 'args', False)
+                                    except KeyError:
+                                        continue
+                                    if not args:
+                                        continue
+                                    if not args[0] == arg:
+                                        continue
+                                target_ty = state.typemap[arg.name]
+                                if not isinstance(target_ty,
+                                                  self._accepted_types):
+                                    continue
+                                tuple_getitem = stmt
+                                break
+                    if tuple_getitem:
+                        break
+                else:
+                    continue  # no getitem in this loop
+
+                ui = unroll_info(loop, literal_unroll_call, arg,
+                                 tuple_getitem)
+                literal_unroll_info.append(ui)
+            return literal_unroll_info
+
+        # 1. Collect info about the literal_unroll loops, ensure they are legal
+        literal_unroll_loops = find_unroll_loops(loops)
```

Comment:
My understanding of ee4cae1cc43613684c74c983e129e08e3d59c600 is that `literal_unroll_info` was previously the dictionary that has the collision. It is now changed into a list to prevent collision.  I agree that `test_22` should have been an unsupported case.

Reply:
@sklam this is correct. The algorithm essentially looked in the blocks reachable by a `literal_unroll` induced loop for _any_ getitem, without considering that a nested `literal_unroll` loop would necessarily have one loop with a getitem on its induction variable and another loop which with no getitem on its induction variable. The result would be that both loops would reference the same getitem, hence collision when used as a dictionary key. Making the case in `test_22` greatly simplifies things without material loss of functionality (there's another way to express the behaviour, i.e. `range(len(<container>))`).